### ISSUE TITLE
Application to MNIST dataset

### ISSUE THREAD

**Issue by bloomingfield** [thwjoy/ccvae#1]
Hi Tom!

This is a really great paper, and a really cool way to apply VAE's. I've just started my own PhD, and I've been working towards reproducing your results from table 6 in the paper, where this method is applied in the context of multi-class classification. I've made some headway, but I can't quite reach the same level of performance as you did! Would you be happy to also share the code that was used for that dataset? If it's easier to just share the pyro code, I can work with that too.

Thanks so much!
Nathaniel

**Comment by thwjoy** [comment]
Hi Nathaniel,

Thank you! To obtain the best results for multi-class classification you
need to enumerate out y, i.e. just sum all of the values rather than
sampling if for unsupervised case. You can do this quire easily in Pyro
using enuemrate
https://pyro.ai/examples/ss-vae.html?highlight=semi%20supervised.  What
sort of accuracy are you getting? I can share the code with you if you
wish, but it's a big mess and I'd rather keep it private if I can.

Best Wishes,
Tom

**Comment by bloomingfield** [comment]
Hi Tom

Thanks for getting back to me so quickly! I'll keep in mind enumerating over the y labels, and no worries at all about your code. Maybe I could ask a few questions?

- Did you change everything to cross-entropy loss (e.g. categorical distributions instead of Bernoulli) for the MNIST data?
- What networks did you use for the MNIST case?
- What hyper-parameters did you use to train the best models?

I also noticed that on line 103 in ccvae.py in the pytorch version of this repo, you use both log_qyzc_ (where Z gradient is detached) and log_qyzc (where Z gradient is not detached). Is this at odds with equation 8 in the paper and the Pyro repo? Or I might have missed something...

I'm currently getting ~85% accuracy for 0.4% labelled, ~96% for 6% labelled, and ~98% for 100% labelled on MNIST. I'm most likely using very different architectures to you though

**Comment by bloomingfield** [comment]
Hi Tom

Thanks for your help with this! I've put together a pyro best guess in a fork

https://github.com/bloomingfield/ccvae

I've just upscaled the MNIST data so that the same networks and hyperparameters can be used as for CELEBA. Are there any additional tricks that were used to get that last bit of accuracy? Particularly in the f=0.004 labelled case, this code seems to struggle (even with enumerating out y).
